src/cli/index.ts

```diff
@@ -89,46 +89,52 @@ export default class Cli {
-            headers.Authorization = `Bearer ${process.env.CUCUMBER_PUBLISH_TOKEN}`
-          }
+    const formatters = await bluebird.map(
+      formats,
+      async ({ type, outputTo }) => {
+        let stream: IFormatterStream = this.stdout
+        if (outputTo !== '') {
+          if (outputTo.match(new RegExp('^https?://')) !== null) {
+            const headers: { [key: string]: string } = {}
+            if (process.env.CUCUMBER_PUBLISH_TOKEN !== undefined) {
+              headers.Authorization = `Bearer ${process.env.CUCUMBER_PUBLISH_TOKEN}`
+            }
 
-          stream = new HttpStream(outputTo, 'GET', headers, (content) =>
-            console.error(content)
+            stream = new HttpStream(outputTo, 'GET', headers, (content) =>
+              console.error(content)
+            )
+          } else {
+            const fd = await fs.open(path.resolve(this.cwd, outputTo), 'w')
+            stream = fs.createWriteStream(null, { fd })
+          }
+          streamsToClose.push(stream)
+        }
+        const typeOptions = {
+          cwd: this.cwd,
+          eventBroadcaster,
+          eventDataCollector,
+          log: stream.write.bind(stream),
+          parsedArgvOptions: formatOptions,
+          stream,
+          supportCodeLibrary,
+        }
+        if (doesNotHaveValue(formatOptions.colorsEnabled)) {
+          typeOptions.parsedArgvOptions.colorsEnabled = (stream as TtyWriteStream).isTTY
+        }
+        if (type === 'progress-bar' && !(stream as TtyWriteStream).isTTY) {
+          const outputToName = outputTo === '' ? 'stdout' : outputTo
+          console.warn(
+            `Cannot use 'progress-bar' formatter for output to '${outputToName}' as not a TTY. Switching to 'progress' formatter.`
           )
-        } else {
-          const fd = await fs.open(path.resolve(this.cwd, outputTo), 'w')
-          stream = fs.createWriteStream(null, { fd })
+          type = 'progress'
         }
-        streamsToClose.push(stream)
+        return FormatterBuilder.build(type, typeOptions)
       }
-      const typeOptions = {
-        cwd: this.cwd,
-        eventBroadcaster,
-        eventDataCollector,
-        log: stream.write.bind(stream),
-        parsedArgvOptions: formatOptions,
-        stream,
-        supportCodeLibrary,
-      }
-      if (doesNotHaveValue(formatOptions.colorsEnabled)) {
-        typeOptions.parsedArgvOptions.colorsEnabled = (stream as TtyWriteStream).isTTY
-      }
-      if (type === 'progress-bar' && !(stream as TtyWriteStream).isTTY) {
-        const outputToName = outputTo === '' ? 'stdout' : outputTo
-        console.warn(
-          `Cannot use 'progress-bar' formatter for output to '${outputToName}' as not a TTY. Switching to 'progress' formatter.`
-        )
-        type = 'progress'
-      }
-      return FormatterBuilder.build(type, typeOptions)
-    })
+    )
     return async function () {
+      await bluebird.each(formatters, async (formatter) => {
```

Comment:
Almost all of the rest of this file is just indentation changes - this is the addition, where we wait for each formatter to be finished.

Reply:
Thoughts on making the default formatter `finished` function close the stream?
Then for the html formatter, it waits for the cucumber html stream to finish and then also awaits on super? Think that simplifies this a little bit.

Reply:
That does seem like a good balance, although there's a special case where if the formatter is writing to `stdout` then we can't end it in the formatter as it'd be too early. I'll try and do something to account for that.

Reply:
If a formatter's writable stream is written to by piping another readable stream into it, then we just have to wait until the pipe is finished. We cannot prematurely close it.

This is what's happening in this case. The [CucumberHtmlStream](https://github.com/cucumber/cucumber/blob/master/html-formatter/javascript/src/CucumberHtmlStream.ts) is piped to the formatter's stream. We have no control over when it's done.

Reply:
Okay I have pushed something which I believe accounts for all of this whilst keeping a straightforward API for formatter authors, let me know what you think. Also clarified my comment above.

Reply:
> That does seem like a good balance.
> 
> In cases where the formatter has asked to write to `stdout`, we can't just end the stream when the formatter is done writing, because the process still needs `stdout`. The current code accounts for this by not adding it to the list of streams to end.

I think it would be best to wait for it to have written everything. The correct way to keep stdout open is to `stream.pipe(process.stdout, {end: false})`